var/spack/repos/builtin/packages/py-smartsim/package.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+from spack.package import *
+
+
+class PySmartsim(PythonPackage):
+    """A workflow library to easily integrate machine learning libraries with
+    high performance computing simulations and applications
+    """
+
+    homepage = "https://www.craylabs.org/docs/overview.html"
+    git = "https://github.com/CrayLabs/SmartSim"
+    pypi = "smartsim/smartsim-0.4.2.tar.gz"
+
+    maintainers("MattToast")
+
+    version("0.5.0", sha256="35b36243dc84af62261a7f772bae92f0b3502faf01401423899cb2a48339858c")
+    version("0.4.2", sha256="ab632ff7d036e73822ddc5081fe85ea69c48d8be53ad0e6e487e9193eb3410f6")
+
+    variant("torch", default=True, description="Build with the pytorch backend")
+    variant("cuda", default=False, description="Use CUDA")
+    variant("rocm", default=False, description="Use ROCm")
```

Comment:
Question for reviewers: I currently have added a cuda/rocm variants to the SmartSim package. These variants do not affect the build process for SmartSim, but they do influence the dependencies that are needed for SmartSim (i.e. which version of Torch to use).

Should this package instead be using the Cuda/Rocm Package mixins instead of manually re-implementing these variants even if the build process of SmartSim specifically is unchanged depending on the variant?

Reply:
IMO, you shouldn't add these variants since they only affect dependencies, not the package itself. You can control variant deps in `packages.yaml` or on the CLI like so:
```console
> spack install py-smartsim ^py-torch +cuda cuda_arch=...
```

Reply:
That makes seems to make sense!

I have one scenario that I'm a little worried about occurring, that I wanted to throw your way to get your two cents:
```sh
> spack install py-samrtsim ^redis-ai ~cuda  ^py-torch +cuda cuda_arch=...
```
The original point of the `cuda`/`rocm` variants was to ensure that RedisAI and the ML deps did not get "out of sync". In the example above, one could trace a model against a gpu using Troch, but would be unable to use it through SmartSim as Redis AI was compiled without gpu support.

If we are willing to chalk this up to simple "user error" I am more than willing to remove the variants.

Reply:
If you explicitly request PyTorch with CUDA support and smartsim without CUDA support, I don't see any issues with giving the user what they ask for. My bigger concern is when this is done implicitly without the user being aware. I'm hoping that our concretizer can get to the point where requesting a root package with CUDA support implies building all dependencies with CUDA support as well, but I don't think we're there yet. Maybe @alalazo has additional thoughts on the proliferation of variant propagation code. This same issue plagues MPI, shared libs, and a lot of other variants. It's frustrating to have to add for-loops over all CUDA archs and enable-disable variants of all deps manually in packages just to ensure a consistent environment.

Reply:
Thank you for the input! It definitely sounds like the `+torch`/`+rocm` variants for SmartSim are an anti-pattern for the Spack ecosystem, and I'm inclined to agree that they should be removed, especially if a similar 'issue' exists in other, more mature packages. Will remove!